### ISSUE TITLE
[no-misused-promises] Array<Promise> as an element of Promise.all

### ISSUE THREAD

**Issue by Cellule** [typescript-eslint/typescript-eslint#2861]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-misused-promises": ["error"]
  }
}
```

```TS
// your repro code case
async function foo(arr: number[]) {
  await Promise.all([
     someOtherAsyncOperation(),
     arr.map(value => asyncSendValue(value))
  ])
}
```

**Expected Result**

The rule should warn that one of the value in `Promise.all` is an `Array<Promise>`

**Actual Result**

No errors

**Additional Info**
I'm not 100% sure if this should be a feature request or a bug report.
It might be appropriate in rule `no-floating-promises` as well.

Basically, one of our devs wrote a similar code as the repro case and it took a long time to figure out where the orphaned promises was.
The fix was to spread `...arr.map(…)` so all promises would be correctly awaited.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.9.1` |
| `@typescript-eslint/parser`        | `4.9.1` |
| `TypeScript`                       | `4.1.2` |
| `ESLint`                           | `7.15.0` |
| `node`                             | `14.11.0` |

**Comment by bradzacher** [comment]
`no-misused-promises` isn't the correct rule for this functionality as there isn't a misused promise (an `Array<Promise>` isn't a promise).

---

This is a very unique case as well.

If you were using the value, then you would easily uncover this bug because the return type of your `Promise.all` will be `Promise<[ReturnType<typeof someOtherAsyncOperation>, Promise<number>[]]>` (so when you iterate the array and you'll immediately run into issues operating on a `Promise<number>[]`).

It's only in this specific use case where `Promise.all` is being used as a synchronisation point and a `void` await that you will run into an issue.

I'm not sure if this rule belongs specifically in this plugin. Whilst it would require type information to reliably detect - there's nothing specific to typescript here. It might be better to include in something like [eslint-plugin-promise](https://github.com/xjamundx/eslint-plugin-promise)?

**Comment by Cellule** [comment]
Yeah I agree that the use case is very specific, although we do use that pattern often-ish... 
I figured I'd give it a shot. I might fiddle to see if I can write a simple rule for that specific case and then it might be easier to understand where it could fit

Thanks for the quick response
